Mathlib/Topology/UniformSpace/MultipliableTendstoUniformly.lean

```diff
@@ -0,0 +1,121 @@
+    (f : ι → α → β)  {p : Filter ι} {g : α → β} {K : Set α} {T V : β} (htv : T < V)
+    (hf : TendstoUniformlyOn f g p K) (hg : ∀ x, x ∈ K → g x ≤ T) :
+    ∀ᶠ (n : ι) in p, ∀ x, x ∈ K → f n x ≤ V := by
+    filter_upwards [hf.eventually_forall_lt htv hg] with i hi x hx using (hi x hx).le
+
+lemma tendstoUniformlyOn_comp_cexp {p : Filter ι} {f : ι → α → ℂ} {g : α → ℂ}
+    {K : Set α} (hf : TendstoUniformlyOn f g p K) (hg : BddAbove ((fun x => (g x).re) '' K)) :
+    TendstoUniformlyOn (fun n => fun x => cexp (f n x)) (cexp ∘ g) p K := by
+  obtain ⟨T, hT⟩ := hg
+  simp only [mem_upperBounds, Set.mem_image, forall_exists_index, and_imp,
+    forall_apply_eq_imp_iff₂] at hT
+  have h2 := TendstoUniformlyOn.eventually_forall_le (fun n x => (f n x).re) (lt_add_one T)
+    hf.re hT
+  have w2 := tendstoUniformlyOn_univ.mpr <| UniformContinuousOn.comp_tendstoUniformly_eventually
+    {x : ℂ | x.re ≤ T + 1} (fun a => K.restrict (f (a))) (fun b => g b) (by simpa using h2) ?_
+      (UniformContinuousOn.cexp (T + 1)) ((tendstouniformlyOn_iff_restrict).mp hf)
+  · rw [tendstouniformlyOn_iff_restrict, ← tendstoUniformlyOn_univ]
+    exact w2
+  · simp only [Set.mem_setOf_eq, Subtype.forall]
+    exact fun a ha => le_trans (hT a ha) (by aesop)
+
+lemma tendstoUniformlyOn_tprod_of_clog {f : ι → α → ℂ} {K : Set α}
+    (h : ∀ x, x ∈ K → Summable fun n => log (f n x))
+    (hf : TendstoUniformlyOn (fun n a => ∑ i ∈ n, log (f i a))
+    (fun a : α => ∑' n : ι, log (f n a)) atTop K) (hfn : ∀ x, x ∈ K → ∀ n : ι, f n x ≠ 0)
+    (hg : BddAbove ((fun x => (∑' n : ι, log (f n x)).re) '' K)) :
+    TendstoUniformlyOn (fun n a => ∏ i ∈ n, (f i a)) (fun a => ∏' i, (f i a)) atTop K := by
+  suffices HU : TendstoUniformlyOn (fun n a => ∏ i ∈ n, (f i a))
+       (cexp ∘ fun a ↦ ∑' n : ι, log (f n a)) atTop K by
+        apply TendstoUniformlyOn.congr_right HU
+        intro x hx
+        simpa using (Complex.cexp_tsum_eq_tprod _ (hfn x hx) (h x hx))
+  apply TendstoUniformlyOn.congr (tendstoUniformlyOn_comp_cexp hf hg)
+  filter_upwards with s i hi using by simp [exp_sum, fun y ↦ exp_log (hfn i hi y)]
+
+/-- This is a version of `tendstoUniformlyOn_tprod_of_clog` for nat that uses range
+in the products. -/
+lemma tendstoUniformlyOn_tprod_nat_of_clog {f : ℕ → α → ℂ} {K : Set α}
+    (h : ∀ x, x ∈ K → Summable fun n => log (f n x))
+    (hf : TendstoUniformlyOn (fun n a => ∑ i ∈ n, log (f i a)) (fun a : α => ∑' n : ℕ, log (f n a))
+    atTop K) (hfn : ∀ x, x ∈ K → ∀ n : ℕ, f n x ≠ 0)
+    (hg : BddAbove ((fun x => (∑' n : ℕ, log (f n x)).re) '' K)) :
+    TendstoUniformlyOn (fun n : ℕ => fun a : α => ∏ i ∈ Finset.range n, (f i a))
+    (fun a => ∏' i, (f i a)) atTop K :=
+  fun v hv => tendsto_finset_range.eventually (tendstoUniformlyOn_tprod_of_clog h hf hfn hg v hv)
+
+/-- This is the version for infinite products of with terms of the form `1 + f n x`. -/
+lemma tendstoUniformlyOn_tprod_nat [TopologicalSpace α] {f : ℕ → α → ℂ} {K : Set α}
```

Comment:
"The version" of what? Please specify. Moreover, why do you require here that the bound on `f` holds for _all_ `n`, not just eventually as in the previous lemma? This complicates the proof while also weakening the statement – you lose both ways.